python_modules/dagster/dagster/_config/structured_config/typing_utils.py

```diff
@@ -0,0 +1,49 @@
+from typing import TYPE_CHECKING, Any, Optional, TypeVar, Union, cast
+
+if TYPE_CHECKING:
+    from dagster._config.structured_config import PartialResource
+
+Self = TypeVar("Self", bound="AllowPartialResourceInitParams")
+
+
+class AllowPartialResourceInitParams:
+    """
+    Implementation of the Python descriptor protocol (https://docs.python.org/3/howto/descriptor.html)
+    to adjust the types of resource inputs and outputs, e.g. resource dependencies can be passed in
+    as PartialResources or Resources, but will always be returned as Resources.
```

Comment:
Please note here that this class exists _only_ to satisfy type-checking libraries. I would also rename the class to reflect that.